File: educational-core/src/com/jetbrains/edu/learning/newproject/CourseProjectGenerator.java
/*
 * Copyright 2000-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jetbrains.edu.learning.newproject;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.annotations.VisibleForTesting;
import com.intellij.ide.RecentProjectsManager;
import com.intellij.ide.util.PropertiesComponent;
import com.intellij.idea.ActionsBundle;
import com.intellij.openapi.application.WriteAction;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.util.Key;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.platform.PlatformProjectOpenProcessor;
import com.intellij.projectImport.ProjectOpenedCallback;
import com.intellij.util.PathUtil;
import com.jetbrains.edu.coursecreator.CCUtils;
import com.jetbrains.edu.coursecreator.yaml.YamlFormatSynchronizer;
import com.jetbrains.edu.learning.EduCourseBuilder;
import com.jetbrains.edu.learning.EduSettings;
import com.jetbrains.edu.learning.EduUtils;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.EduCourse;
import com.jetbrains.edu.learning.courseFormat.Lesson;
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils;
import com.jetbrains.edu.learning.statistics.EduCounterUsageCollector;
import com.jetbrains.edu.learning.stepik.StepikNames;
import com.jetbrains.edu.learning.stepik.StepikSolutionsLoader;
import com.jetbrains.edu.learning.stepik.StepikUser;
import com.jetbrains.edu.learning.stepik.api.StepikConnector;
import com.jetbrains.edu.learning.stepik.api.StepikCourseLoader;
import com.jetbrains.edu.learning.yaml.EduYamlUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.util.EnumSet;

public abstract class CourseProjectGenerator<S> {

  public static final Key<Boolean> EDU_PROJECT_CREATED = Key.create("edu.projectCreated");
  public static final Key<String> COURSE_MODE_TO_CREATE = Key.create("edu.courseModeToCreate");
  public static final Key<String> COURSE_TYPE_TO_CREATE = Key.create("edu.courseTypeToCreate");
  public static final Key<String> COURSE_LANGUAGE_ID_TO_CREATE = Key.create("edu.courseLanguageIdToCreate");

  private static final Logger LOG = Logger.getInstance(CourseProjectGenerator.class);

  @NotNull protected final EduCourseBuilder<S> myCourseBuilder;
  @NotNull protected Course myCourse;
  private boolean alreadyEnrolled;

  public CourseProjectGenerator(@NotNull EduCourseBuilder<S> builder, @NotNull final Course course) {
    myCourseBuilder = builder;
    myCourse = course;
  }

  protected boolean beforeProjectGenerated() {
    if (!(myCourse instanceof EduCourse) || !((EduCourse)myCourse).isRemote()) return true;
    final EduCourse remoteCourse = (EduCourse) this.myCourse;
    if (remoteCourse.getId() > 0) {
      return ProgressManager.getInstance().runProcessWithProgressSynchronously(() -> {
        ProgressManager.getInstance().getProgressIndicator().setIndeterminate(true);

        final StepikUser user = EduSettings.getInstance().getUser();
        if (user != null) {
          alreadyEnrolled = StepikConnector.getInstance().isEnrolledToCourse(remoteCourse.getId(), user);
          if (!alreadyEnrolled) {
            StepikConnector.getInstance().enrollToCourse(remoteCourse.getId(), user);
          }
        }
        StepikCourseLoader.loadCourseStructure(remoteCourse);
        myCourse = remoteCourse;
        return true;
      }, "Loading Course", true, null);
    }
    return true;
  }

  protected void afterProjectGenerated(@NotNull Project project, @NotNull S projectSettings) {
    loadSolutions(project, myCourse);
    EduUtils.openFirstTask(myCourse, project);
    boolean courseCreator = CCUtils.isCourseCreator(project);
    ObjectMapper mapper = courseCreator ? YamlFormatSynchronizer.getMAPPER() : EduYamlUtil.getEDU_MAPPER();
    YamlFormatSynchronizer.saveAll(project, mapper);
    YamlFormatSynchronizer.startSynchronization(project);
  }

  /**
   * Generate new project and create course structure for created project
   *
   * @param location location of new course project
   * @param projectSettings new project settings
   */
  // 'projectSettings' must have S type but due to some reasons:
  //  * We don't know generic parameter of EduPluginConfigurator after it was gotten through extension point mechanism
  //  * Kotlin and Java do type erasure a little bit differently
  // we use Object instead of S and cast to S when it needed
  @SuppressWarnings("unchecked")
  @Nullable
  public final Project doCreateCourseProject(@NotNull String location, @NotNull Object projectSettings) {
    if (!beforeProjectGenerated()) {
      return null;
    }
    S castedProjectSettings = (S) projectSettings;
    Project createdProject = createProject(location, castedProjectSettings);
    if (createdProject == null) return null;
    afterProjectGenerated(createdProject, castedProjectSettings);
    return createdProject;
  }

  /**
   * Create new project in given location.
   * To create course structure: modules, folders, files, etc. use {@link CourseProjectGenerator#createCourseStructure(Project, VirtualFile, Object)}
   *
   * @param locationString location of new project
   * @param projectSettings new project settings
   * @return project of new course or null if new project can't be created
   */
  @Nullable
  protected Project createProject(@NotNull String locationString, @NotNull S projectSettings) {
    final File location = new File(FileUtil.toSystemDependentName(locationString));
    if (!location.exists() && !location.mkdirs()) {
      String message = ActionsBundle.message("action.NewDirectoryProject.cannot.create.dir", location.getAbsolutePath());
      Messages.showErrorDialog(message, ActionsBundle.message("action.NewDirectoryProject.title"));
      return null;
    }

    final VirtualFile baseDir = WriteAction.compute(() -> LocalFileSystem.getInstance().refreshAndFindFileByIoFile(location));
    if (baseDir == null) {
      LOG.error("Couldn't find '" + location + "' in VFS");
      return null;
    }
    VfsUtil.markDirtyAndRefresh(false, true, true, baseDir);

    RecentProjectsManager.getInstance().setLastProjectCreationLocation(PathUtil.toSystemIndependentName(location.getParent()));

    ProjectOpenedCallback callback = (p, module) -> createCourseStructure(p, baseDir, projectSettings);
    EnumSet<PlatformProjectOpenProcessor.Option> options = EnumSet.of(PlatformProjectOpenProcessor.Option.FORCE_NEW_FRAME);
    baseDir.putUserData(COURSE_MODE_TO_CREATE, myCourse.getCourseMode());
    baseDir.putUserData(COURSE_TYPE_TO_CREATE, myCourse.getItemType());
    baseDir.putUserData(COURSE_LANGUAGE_ID_TO_CREATE, myCourse.getLanguageID());
    Project project = PlatformProjectOpenProcessor.doOpenProject(baseDir, null, -1, callback, options);
    if (project != null) {
      project.putUserData(EDU_PROJECT_CREATED, true);
    }
    return project;
  }

  /**
   * Create course structure for already created project.
   *
   * @param project course project
   * @param baseDir base directory of project
   * @param settings project settings
   */
  @VisibleForTesting
  public void createCourseStructure(@NotNull Project project, @NotNull VirtualFile baseDir, @NotNull S settings) {
    GeneratorUtils.initializeCourse(project, myCourse);

    if (CCUtils.isCourseCreator(project) && myCourse.getItems().isEmpty()) {
      final Lesson lesson = myCourseBuilder.createInitialLesson(project, myCourse);
      if (lesson != null) {
        myCourse.addLesson(lesson);
      }
    }

    try {
      ProgressManager.getInstance().runProcessWithProgressSynchronously(() -> {
        ProgressIndicator indicator = ProgressManager.getInstance().getProgressIndicator();
        GeneratorUtils.createCourse(myCourse, baseDir, indicator);
        if (CCUtils.isCourseCreator(project)) {
          CCUtils.initializeCCPlaceholders(project, myCourse);
        }
        if (myCourse instanceof EduCourse && ((EduCourse)myCourse).isRemote() && CCUtils.isCourseCreator(project)) {
          checkIfAvailableOnRemote();
        }
        createAdditionalFiles(project, baseDir);
        EduCounterUsageCollector.eduProjectCreated(myCourse.getCourseMode());

        return null; // just to use correct overloading of `runProcessWithProgressSynchronously` method
      }, "Generating Course Structure", false, project);
    } catch (IOException e) {
      LOG.error("Failed to generate course", e);
    }
  }

  private void checkIfAvailableOnRemote() {
    EduCourse courseFromStepik = StepikConnector.getInstance().getCourseInfo(myCourse.getId(), null, true);
    if (courseFromStepik == null) {
      LOG.warn("Failed to get stepik course for imported from zip course with id: " + myCourse.getId());
      LOG.info("Converting course to local. Course id: " + myCourse.getId());
      ((EduCourse)myCourse).convertToLocal();
    }
  }

  protected void loadSolutions(@NotNull Project project, @NotNull Course course) {
    if (course.isStudy() && course instanceof EduCourse && ((EduCourse)course).isRemote() && EduSettings.isLoggedIn()) {
      PropertiesComponent.getInstance(project).setValue(StepikNames.ARE_SOLUTIONS_UPDATED_PROPERTY, true, false);
      if (alreadyEnrolled) {
        StepikSolutionsLoader stepikSolutionsLoader = StepikSolutionsLoader.getInstance(project);
        stepikSolutionsLoader.loadSolutionsInBackground();
        EduCounterUsageCollector.synchronizeCourse(EduCounterUsageCollector.SynchronizeCoursePlace.PROJECT_GENERATION);
      }
    }
  }

  /**
   * Creates additional files that are not in course object
   *
   * @param project course project
   * @param baseDir base directory of project
   *
   * @throws IOException
   */
  protected void createAdditionalFiles(@NotNull Project project, @NotNull VirtualFile baseDir) throws IOException {}
}

File: educational-core/src/com/jetbrains/edu/learning/newproject/CourseProjectGenerator.kt
package com.jetbrains.edu.learning.newproject

import com.google.common.annotations.VisibleForTesting
import com.intellij.ide.RecentProjectsManager
import com.intellij.ide.impl.OpenProjectTask
import com.intellij.ide.impl.TrustedPaths
import com.intellij.idea.ActionsBundle
import com.intellij.openapi.application.WriteAction
import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.module.Module
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.NOTIFICATIONS_SILENT_MODE
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.ex.ProjectManagerEx
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.util.Key
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetsManager
import com.intellij.profile.codeInspection.ProjectInspectionProfileManager
import com.intellij.util.PathUtil
import com.jetbrains.edu.coursecreator.CCUtils
import com.jetbrains.edu.coursecreator.CCUtils.isLocalCourse
import com.jetbrains.edu.learning.*
import com.jetbrains.edu.learning.courseFormat.Course
import com.jetbrains.edu.learning.courseFormat.CourseMode
import com.jetbrains.edu.learning.courseFormat.CourseVisibility.FeaturedVisibility
import com.jetbrains.edu.learning.courseFormat.EduCourse
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils.IdeaDirectoryUnpackMode.ONLY_IDEA_DIRECTORY
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils.unpackAdditionalFiles
import com.jetbrains.edu.learning.marketplace.MARKETPLACE
import com.jetbrains.edu.learning.marketplace.api.MarketplaceConnector
import com.jetbrains.edu.learning.messages.EduCoreBundle
import com.jetbrains.edu.learning.navigation.NavigationUtils
import com.jetbrains.edu.learning.statistics.EduCounterUsageCollector
import com.jetbrains.edu.learning.stepik.hyperskill.courseGeneration.HyperskillCourseProjectGenerator
import com.jetbrains.edu.learning.yaml.YamlFormatSynchronizer
import java.io.File
import java.io.IOException
import java.nio.file.Path
import kotlin.system.measureTimeMillis

/**
 * If you add any new public methods here, please do not forget to add it also to
 * @see HyperskillCourseProjectGenerator
 */
abstract class CourseProjectGenerator<S : EduProjectSettings>(
  protected val courseBuilder: EduCourseBuilder<S>,
  protected var course: Course
) {
  private var alreadyEnrolled = false

  open fun afterProjectGenerated(project: Project, projectSettings: S) {
    // project.isLocalCourse info is stored in PropertiesComponent to keep it after course restart on purpose
    // not to show login widget for local course
    project.isLocalCourse = course.isLocal

    val statusBarWidgetsManager = project.service<StatusBarWidgetsManager>()
    statusBarWidgetsManager.updateAllWidgets()

    setUpPluginDependencies(project, course)

    NavigationUtils.openFirstTask(course, project)

    YamlFormatSynchronizer.saveAll(project)
    YamlFormatSynchronizer.startSynchronization(project)
  }

  // 'projectSettings' must have S type but due to some reasons:
  //  * We don't know generic parameter of EduPluginConfigurator after it was gotten through extension point mechanism
  //  * Kotlin and Java do type erasure a little differently
  // we use Object instead of S and cast to S when it needed
  fun doCreateCourseProject(location: String, projectSettings: EduProjectSettings): Project? {
    val createdProject = createProject(location) ?: return null

    @Suppress("UNCHECKED_CAST")
    val castedProjectSettings = projectSettings as S
    afterProjectGenerated(createdProject, castedProjectSettings)
    return createdProject
  }

  /**
   * Create new project in given location.
   * To create course structure: modules, folders, files, etc. use [CourseProjectGenerator.createCourseStructure]
   *
   * @param locationString location of new project
   *
   * @return project of new course or null if new project can't be created
   */
  private fun createProject(locationString: String): Project? {
    val location = File(FileUtil.toSystemDependentName(locationString))
    if (!location.exists() && !location.mkdirs()) {
      val message = ActionsBundle.message("action.NewDirectoryProject.cannot.create.dir", location.absolutePath)
      Messages.showErrorDialog(message, ActionsBundle.message("action.NewDirectoryProject.title"))
      return null
    }
    val baseDir = WriteAction.compute<VirtualFile?, RuntimeException> { LocalFileSystem.getInstance().refreshAndFindFileByIoFile(location) }
    if (baseDir == null) {
      LOG.error("Couldn't find '$location' in VFS")
      return null
    }
    VfsUtil.markDirtyAndRefresh(false, true, true, baseDir)

    RecentProjectsManager.getInstance().lastProjectCreationLocation = PathUtil.toSystemIndependentName(location.parent)

    baseDir.putUserData(COURSE_MODE_TO_CREATE, course.courseMode)
    baseDir.putUserData(COURSE_LANGUAGE_ID_TO_CREATE, course.languageId)

    if (isCourseTrusted(course, isNewCourseCreatorCourse)) {
      @Suppress("UnstableApiUsage")
      TrustedPaths.getInstance().setProjectPathTrusted(location.toPath(), true)
    }

    val holder = CourseInfoHolder.fromCourse(course, baseDir)
    // @formatter:off
    ProgressManager.getInstance().runProcessWithProgressSynchronously<Unit, IOException>({
      createCourseStructure(holder)
    }, EduCoreBundle.message("generate.project.generate.course.structure.progress.text"), false, null)
    // @formatter:on

    val newProject = openNewCourseProject(course, location.toPath(), this::prepareToOpen) ?: return null

    // @formatter:off
    ProgressManager.getInstance().runProcessWithProgressSynchronously<Unit, IOException>({
      unpackAdditionalFiles(holder, ONLY_IDEA_DIRECTORY)
    }, EduCoreBundle.message("generate.project.unpack.course.project.settings.progress.text"), false, newProject)
    // @formatter:on

    // after adding files with settings to .idea directory, almost all settings are synchronized automatically,
    // but the inspection profiles are to be synchronized manually
    ProjectInspectionProfileManager.getInstance(newProject).initializeComponent()

    return newProject
  }

  protected open suspend fun prepareToOpen(project: Project, module: Module) {
    NOTIFICATIONS_SILENT_MODE.set(project, true)
  }

  private fun openNewCourseProject(
    course: Course,
    location: Path,
    prepareToOpenCallback: suspend (Project, Module) -> Unit
  ): Project? {
    val task = OpenProjectTask(course, prepareToOpenCallback)

    return ProjectManagerEx.getInstanceEx().openProject(location, task)
  }

  private fun OpenProjectTask(course: Course, prepareToOpenCallback: suspend (Project, Module) -> Unit): OpenProjectTask {
    @Suppress("UnstableApiUsage")
    return OpenProjectTask {
      forceOpenInNewFrame = true
      isNewProject = true
      isProjectCreatedWithWizard = false
      runConfigurators = true
      beforeInit = {
        it.putUserData(EDU_PROJECT_CREATED, true)
      }
      preparedToOpen = {
        StudyTaskManager.getInstance(it.project).course = course
        prepareToOpenCallback(it.project, it)
      }
    }
  }

  /**
   * Creates course structure in directory provided by [holder]
   */
  @VisibleForTesting
  open fun createCourseStructure(holder: CourseInfoHolder<Course>) {
    holder.course.init(false)
    val isNewCourseCreatorCourse = isNewCourseCreatorCourse

    if (isNewCourseCreatorCourse) {
      val lesson = courseBuilder.createInitialLesson(holder)
      if (lesson != null) {
        course.addLesson(lesson)
      }
    }

    try {
      generateCourseContent(holder, isNewCourseCreatorCourse, ProgressManager.getInstance().progressIndicator)
    }
    catch (e: IOException) {
      LOG.error("Failed to generate course", e)
    }
  }

  @Throws(IOException::class)
  private fun generateCourseContent(
    holder: CourseInfoHolder<Course>,
    isNewCourseCreatorCourse: Boolean,
    indicator: ProgressIndicator
  ) {
    val duration = measureTimeMillis {
      val course = holder.course
      if (!course.isStudy) {
        CCUtils.initializeCCPlaceholders(holder)
      }
      GeneratorUtils.createCourse(holder, indicator)
      if (course is EduCourse && course.isMarketplaceRemote && !course.isStudy) {
        checkIfAvailableOnRemote(course)
      }
      createAdditionalFiles(holder, isNewCourseCreatorCourse)
      EduCounterUsageCollector.eduProjectCreated(course)
    }
    LOG.info("Course content generation: $duration ms")
  }

  private fun checkIfAvailableOnRemote(course: EduCourse) {
    val remoteCourse = MarketplaceConnector.getInstance().searchCourse(course.id, course.isMarketplacePrivate)
    if (remoteCourse == null) {
      LOG.warn("Failed to get $MARKETPLACE course for imported from zip course with id: ${course.id}")
      LOG.info("Converting course to local. Course id: ${course.id}")
      course.convertToLocal()
    }
  }

  /**
   * Creates additional files that are not in course object
   *
   * @param holder contains info about course project like root directory
   * @param isNewCourse `true` if course is new one, `false` otherwise
   *
   * @throws IOException
   */
  @Throws(IOException::class)
  open fun createAdditionalFiles(holder: CourseInfoHolder<Course>, isNewCourse: Boolean) {
  }

  private val isNewCourseCreatorCourse: Boolean
    get() = course.courseMode == CourseMode.EDUCATOR && course.items.isEmpty()

  companion object {
    private val LOG: Logger = Logger.getInstance(CourseProjectGenerator::class.java)
    private const val JETBRAINS_SRO_ORGANIZATION: String = "JetBrains s.r.o."

    val EDU_PROJECT_CREATED = Key.create<Boolean>("edu.projectCreated")

    val COURSE_MODE_TO_CREATE = Key.create<CourseMode>("edu.courseModeToCreate")
    val COURSE_LANGUAGE_ID_TO_CREATE = Key.create<String>("edu.courseLanguageIdToCreate")


    // TODO: provide more precise heuristic for Gradle, sbt and other "dangerous" build systems
    // See https://youtrack.jetbrains.com/issue/EDU-4182
    private fun isCourseTrusted(course: Course, isNewCourseCreatorCourse: Boolean): Boolean {
      if (isNewCourseCreatorCourse) return true
      if (course !is EduCourse) return true
      if (course.visibility is FeaturedVisibility) return true
      // Trust any marketplace course from JetBrains s.r.o. organization
      if (course.isMarketplaceRemote && course.organization == JETBRAINS_SRO_ORGANIZATION) return true
      return course.isPreview
    }
  }
}
